src/centralized_telescope.rs

```diff
@@ -0,0 +1,419 @@
+        let (h, h_u64_opt) = Self::h1(r.h.clone().as_ref(), s.as_ref(), r.n_p, sec_param);
+        h_u64_opt.map(|h_u64| Self {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h,
+            h_u64,
+            n_p: r.n_p,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    v: u64,
+    /// Proof 2nd counter
+    t: u64,
+    /// Proof tuple
+    items: Vec<Element>,
+}
+
+impl Proof {
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    fn h0(setup: &Setup, v: u64, s: Element) -> Option<u64> {
+        let mut hasher = Blake2s256::new();
+        hasher.update(b"Telescope-H0");
+        hasher.update(v.to_be_bytes());
+        hasher.update(s);
+        let digest: Hash = hasher.finalize().into();
+        utils::sample_uniform(&digest, setup.n_p, setup.sec_param)
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut hasher = Blake2s256::new();
+        hasher.update(b"Telescope-H2");
+        hasher.update(r.h);
+        let digest: Hash = hasher.finalize().into();
+        utils::sample_bernouilli(&digest, setup.q, setup.sec_param)
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(setup: &Setup, bins: &[Vec<Element>], round: &Round, limit: u64) -> (u64, Option<Self>) {
+        if round.s_list.len() as u64 == setup.u {
+            let proof_opt = if Self::h2(setup, round) {
+                Some(Self {
+                    v: round.v,
+                    t: round.t,
+                    items: round.s_list.clone(),
+                })
+            } else {
+                None
+            };
+            return (limit, proof_opt);
+        }
+
+        let mut l = limit;
+        for &s in &bins[round.h_u64 as usize] {
+            if let Some(r) = Round::update(round, s, setup.sec_param) {
+                let (l_dfs, proof_opt) = Self::dfs(setup, bins, &r, l.saturating_add(1));
+                if proof_opt.is_some() {
+                    return (l_dfs, proof_opt);
+                }
+                l = l_dfs;
+            }
+        }
+        (l, None)
+    }
+
+    /// Indexed proving algorithm, returns an empty proof if no suitable
+    /// candidate is found within the setup.b steps.
+    fn prove_index(setup: &Setup, set: &[Element], v: u64) -> (u64, Option<Self>) {
+        let mut bins: Vec<Vec<Element>> = vec![vec![]; setup.n_p as usize];
+        for &s in set {
+            match Self::h0(setup, v, s) {
+                Some(h) => {
+                    bins[h as usize].push(s);
+                }
+                None => return (0, None),
```

Comment:
Future reminder: When we implement the `errors`, we definitely should specify in which case we return `none`.
Also, it could be useful to explain the difference between returning an empty proof or `none` in the docs.

Reply:
Sure we can be more verbose when using errors (here it is the initialisation that failed for instance, not the proof generation per se).
We do not have empty proofs anymore by the way!

Reply:
Ooops! Yes, I missed it. But, agreed on being more verbose